src/topology/algebra/module/locally_convex.lean

```diff
@@ -0,0 +1,138 @@
+
+lemma locally_convex_of_bases {ι : Type*} (b : E → ι → set E) (p : ι → Prop)
+  (hbasis : ∀ x : E, (𝓝 x).has_basis p (b x)) (hconvex : ∀ x i, p i → convex 𝕂 (b x i)) :
+  locally_convex_space 𝕂 E :=
+⟨λ x, (hbasis x).to_has_basis
+  (λ i hi, ⟨b x i, ⟨⟨(hbasis x).mem_of_mem hi, hconvex x i hi⟩, le_refl (b x i)⟩⟩)
+  (λ s hs, ⟨(hbasis x).index s hs.1,
+    ⟨(hbasis x).property_index hs.1, (hbasis x).set_index_subset hs.1⟩⟩)⟩
+
+lemma locally_convex_space.convex_basis_zero [locally_convex_space 𝕂 E] :
+  (𝓝 0 : filter E).has_basis (λ s, s ∈ (𝓝 0 : filter E) ∧ convex 𝕂 s) id :=
+locally_convex_space.convex_basis 0
+
+lemma locally_convex_iff_exists_convex_subset :
+  locally_convex_space 𝕂 E ↔ ∀ x : E, ∀ U ∈ 𝓝 x, ∃ S ∈ 𝓝 x, convex 𝕂 S ∧ S ⊆ U :=
+begin
+  rw locally_convex_space_iff,
+  refine forall_congr (λ x, _),
+  simp_rw [has_basis_iff, exists_prop, and_assoc],
+  exact ⟨λ H U, (H U).mp, λ H U, ⟨H U, λ ⟨S, hS, _, hS'⟩, mem_of_superset hS hS'⟩⟩
+end
+
+end semimodule
+
+section module
+
+variables (𝕂 E : Type*) [ordered_semiring 𝕂] [add_comm_group E] [module 𝕂 E] [topological_space E]
+  [topological_add_group E]
+
+lemma locally_convex_of_basis_zero {ι : Type*} (b : ι → set E) (p : ι → Prop)
+  (hbasis : (𝓝 0).has_basis p b) (hconvex : ∀ i, p i → convex 𝕂 (b i)) :
+  locally_convex_space 𝕂 E :=
+begin
+  refine locally_convex_of_bases 𝕂 E (λ (x : E) (i : ι), ((+) x) '' b i) p (λ x, _)
+    (λ x i hi, (hconvex i hi).translate x),
+  rw ← map_add_left_nhds_zero,
+  exact hbasis.map _
+end
+
+lemma locally_convex_space_iff_zero :
+  locally_convex_space 𝕂 E ↔
+  (𝓝 0 : filter E).has_basis (λ (s : set E), s ∈ (𝓝 0 : filter E) ∧ convex 𝕂 s) id :=
+⟨λ h, @locally_convex_space.convex_basis _ _ _ _ _ _ h 0,
+ λ h, locally_convex_of_basis_zero 𝕂 E _ _ h (λ s, and.right)⟩
+
+lemma locally_convex_iff_exists_convex_subset_zero :
+  locally_convex_space 𝕂 E ↔
+  ∀ U ∈ (𝓝 0 : filter E), ∃ S ∈ (𝓝 0 : filter E), convex 𝕂 S ∧ S ⊆ U :=
+begin
+  rw locally_convex_space_iff_zero,
+  simp_rw [has_basis_iff, exists_prop, and_assoc],
+  exact ⟨λ H U, (H U).mp, λ H U, ⟨H U, λ ⟨S, hS, _, hS'⟩, mem_of_superset hS hS'⟩⟩
+end
+
+end module
+
+-- TODO : generalize the following section to normed and ordered rings/fields, once we have the
+-- corresponding typeclasses
+
+namespace seminorm
+
+variables {ι 𝕂 E : Type*} [add_comm_group E] [module ℝ E] [nonempty ι] [topological_space E]
+  [topological_add_group E]
+
+lemma locally_convex_space (p : ι → seminorm ℝ E) [with_seminorms p] :
+  locally_convex_space ℝ E :=
+locally_convex_of_basis_zero ℝ E id (λ s, s ∈ seminorm_basis_zero p)
+begin
+  rw [with_seminorms_eq p, add_group_filter_basis.nhds_eq _, add_group_filter_basis.N_zero],
+  exact filter_basis.has_basis _
+end
+begin
+  intros s hs,
+  change s ∈ set.Union _ at hs,
+  simp_rw [set.mem_Union, set.mem_singleton_iff] at hs,
+  rcases hs with ⟨I, r, hr, rfl⟩,
+  exact convex_ball _ _ _
+end
+
+end seminorm
```

Comment:
I am not sure how the import hierarchy should be. My feeling is that this part should be moved to `analysis/seminorm.lean` and that that file should import this one. @YaelDillies what are your thoughts one this?

Reply:
I think `analysis.seminorm` is too big now (congratulations! 🥳) and we should split it into
* the very basics which will go before normed spaces
* the Minkowski functional
* family of seminorms and co

That instance would belong in this last file but not the first. So I'm very happy to leave this instance here until the refactor happens.

Reply:
Indeed this does seem backward. Unless there's already an ongoing refactor, I think I'll move it already otherwise it'll be forgotten.